core/src/main/java/cucumber/runtime/formatter/TimelineFormatter.java

```diff
@@ -0,0 +1,265 @@
+        "/cucumber/formatter/timeline/chosen-sprite.png"
+    };
+
+    private final EventHandler<TestSourceRead> testSourceReadHandler = new EventHandler<TestSourceRead>() {
+        @Override
+        public void receive(TestSourceRead event) {
+            testSources.addTestSourceReadEvent(event.uri, event);
+        }
+    };
+    private final EventHandler<TestCaseStarted> caseStartedHandler = new EventHandler<TestCaseStarted>() {
+        @Override
+        public void receive(TestCaseStarted event) {
+            handleTestCaseStarted(event);
+        }
+    };
+    private final EventHandler<TestCaseFinished> caseFinishedHandler = new EventHandler<TestCaseFinished>() {
+        @Override
+        public void receive(final TestCaseFinished event) {
+            handleTestCaseFinished(event);
+        }
+    };
+    private final EventHandler<TestRunFinished> runFinishedHandler = new EventHandler<TestRunFinished>() {
+        @Override
+        public void receive(final TestRunFinished event) {
+            finishReport(event);
+        }
+    };
+
+    private final TestSourcesModel testSources = new TestSourcesModel();
+    private final List<TestData> allTests = new LinkedList<TestData>();
+    private final Map<Long, GroupData> allGroups = new ConcurrentHashMap<Long, GroupData>();
+    private final URL reportDir;
+    private final NiceAppendable reportJs;
+
+    private TestData currentTest;
+
+    public TimelineFormatter(final URL reportDir) {
+        this(reportDir, createJsonOut(reportDir, "report.js"));
+    }
+
+    TimelineFormatter(final URL reportDir, final NiceAppendable reportJs) {
+        this.reportDir = reportDir;
+        this.reportJs = reportJs;
+    }
+
+    @Override
+    public void setEventPublisher(final EventPublisher publisher) {
+        publisher.registerHandlerFor(TestSourceRead.class, testSourceReadHandler);
+        publisher.registerHandlerFor(TestCaseStarted.class, caseStartedHandler);
+        publisher.registerHandlerFor(TestCaseFinished.class, caseFinishedHandler);
+        publisher.registerHandlerFor(TestRunFinished.class, runFinishedHandler);
+    }
+
+    private void handleTestCaseStarted(final TestCaseStarted event) {
+        final Long threadId = Thread.currentThread().getId();
+        final TestData test = new TestData(event, threadId);
+        currentTest = test;
+        allTests.add(test);
+        if (!allGroups.containsKey(threadId)) {
+            allGroups.put(threadId, new GroupData(threadId));
+        }
+    }
+
+    private void handleTestCaseFinished(final TestCaseFinished event) {
+        currentTest.end(event);
+    }
+
+    private void finishReport(final TestRunFinished event) {
+        final Gson gson = new GsonBuilder().setPrettyPrinting().create();
+        reportJs.append("$(document).ready(function() {");
+        reportJs.println();
+        //Sort results into feature, so can guarantee output of order
+        Collections.sort(allTests, TEST_DATA_COMPARATOR);
```

Comment:
don't actually need to order the output, would just need to update test to ignore order